feat: refactor `OpCodes` API

DavePearce opened the pull request:
Previously, the OpCodes API was implemented via a number of static methods and, once a given fork was loaded, it was not possible to load another fork (doing so would cause a runtime exception).

Therefore, this refactors the OpCodes API to remove entirely the static API and, instead, replace it with an API based around OpCodes *instances*.  That is, we have an instance of OpCodes for each fork under consideration.  This means that, in situations where previously we could look up OpCodeData information statically, we must now be able to access a relevant instance of OpCodes.  As such, quite a lot of refactoring was required to make this possible.

The interface now promoters OpCodeData over OpCode as the primary source of useful information.  The OpCode class is now largely relegated to being simply a "mnemonic" with essentially no actual functionality.  In most cases, this works out well because OpCodeData already had a mnemonic accessor, meaning whereever we have an instance of OpCodeData we also have an instance of OpCode.

DavePearce committed:
refactor OpCodes API to be fork agnostic

Previously, the OpCodes API was implemented via a number of static
methods and, once a given fork was loaded, it was not possible to load
another fork (doing so would cause a runtime exception).

Therefore, this refactors the OpCodes API to remove entirely the static
API and, instead, replace it with an API based around OpCodes
*instances*.  That is, we have an instance of OpCodes for each fork
under consideration.  This means that, in situations where previously we
could look up OpCodeData information statically, we must now be able to
access a relevant instance of OpCodes.  As such, quite a lot of
refactoring was required to make this possible.

The interface now promoters OpCodeData over OpCode as the primary source
of useful information.  The OpCode class is now largely relegated to
being simply a "mnemonic" with essentially no actual functionality.  In
most cases, this works out well because OpCodeData already had a
mnemonic accessor, meaning whereever we have an instance of OpCodeData
we also have an instance of OpCode.

DavePearce committed:
fix OpCodeData.equals => OpCode.equals

DavePearce committed:
fix StpCall

The StpCall class was incorrectly using an OpCodeData field instead of
an OpCode field.  Specifically, this was being used within its equals
and hashCode methods.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -1101,4 +1120,9 @@ protected abstract void traceSystemInitialTransaction(
   protected abstract void setSelfdestructSection(Hub hub, final MessageFrame frame);
+
+  /** Provides access to fork-specific opcode information. */
+  public OpCodes getOpcodes() {
```
better uing a @Getter annotation no ?

DavePearce replied:
Well, if you like Lombok I suppose so ;)

DavePearce replied:
I'll make this change, no probs

DavePearce commented on arithmetization/src/main/java/net/consensys/linea/zktracer/opcode/OpCodes.java:
```diff
@@ -40,74 +40,46 @@
+      opCodesMap.put(fork, instance);
     }
+    // Done
+    return instance;
```
That behaviour is expected and is preferable to requiring an explicit lock.

DavePearce committed:
use @Getter instead of getOpcodes()

DavePearce committed:
fix: BytecodeCompiler.op()

This function previously threw an exception if the opcode provided was
invalid for the given fork.

DavePearce committed:
Use @BeforeAll in conjunction with @BeforeEach

The issue with using @BeforeEach to set the fork and load the opcodes is
that this is not guaranteed to run before methods used as MethodSource's
for parameterised test.  However, @BeforeAll is guaranteed to run before
any test, including any MethodSource methods.  Therefore, splitting up
the process of initialising the TracerTestBase gives the best solution.
